<a href="https://colab.research.google.com/github/Lou-Pomponio/Ejercicios/blob/main/ProyectoFinalLourdesPomponio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**<h1> PROYECTO FINAL LOURDES POMPONIO </h1>**

# LIBRERIAS

In [1]:
#Se instalan e importan las librerias a utilizar

!pip install pandas==2.2.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.0/13.0 MB 65.7 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.0.3
    Uninstalling pandas-2.0.3:
      Successfully uninstalled pandas-2.0.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.0.3, but you have pandas 2.2.2 which is incompatible.


In [2]:
import pandas as pd
import os

# DATASETS ORIGINALES

Se importan los datasets originales que contienen la información a analizar.
Dicha información se encuentra documentada en archivos .csv

In [3]:
#Los archivos csv serán importados como DataFrame.
#Se ordenará ver los primeros 5 registros para comprobar su correcta ejecución.

Datasets: ORDERS



In [4]:
df_orders = pd.read_csv("https://raw.githubusercontent.com/Lou-Pomponio/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

Datasets: ORDER_ITEMS

In [5]:
df_order_items = pd.read_csv("https://raw.githubusercontent.com/Lou-Pomponio/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

Datasets: PRODUCTS

In [6]:
df_products = pd.read_csv("https://raw.githubusercontent.com/Lou-Pomponio/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

Datasets: CUSTOMERS

In [7]:
df_customers = pd.read_csv("https://raw.githubusercontent.com/Lou-Pomponio/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

Datasets: ORDER_PAYMENTS

In [8]:
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/Lou-Pomponio/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

# ESTUDIO DEL DATAFRAME

Se realiza un análisis de cada DataFrame creado para evaluar el tipo de datos que contiene, tamaño y cantidad de nulos.

In [9]:
#Se crea una función que nos permita realizar en cada DF el análisis.
#Luego se aplica en cada DF.
def datosDataFrames(df):
  print("^^^^^^^^^^^^^^")
  print("Tipo de dato")
  print(df.info()) #Tipo de dato del dataframe
  print("^^^^^^^^^^^^^^")
  print("^^^^^^^^^^^^^^")
  print("Tamaño")
  print(df.shape) #Tamaño del dataframe
  print("^^^^^^^^^^^^^^")
  print("^^^^^^^^^^^^^^")
  print("Cantidad de nulos")
  print(df.isnull().sum()) #Recuento de nulos
  print("^^^^^^^^^^^^^^")
  print("^^^^^^^^^^^^^^")
  print("Recuento de duplicados")
  print(df.duplicated().sum()) #Analisis duplicados

In [10]:
datosDataFrames(df_orders)

^^^^^^^^^^^^^^
Tipo de dato
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Tamaño
(99441, 8)
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Cantidad de nulos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_time

In [11]:
datosDataFrames(df_order_items)

^^^^^^^^^^^^^^
Tipo de dato
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Tamaño
(112650, 7)
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Cantidad de nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Recuento de duplicado

In [12]:
datosDataFrames(df_products)

^^^^^^^^^^^^^^
Tipo de dato
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Tamaño
(32951, 9)
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Cantidad de nulos
product_id                      0
product_category_name         610
product_name_len

In [13]:
datosDataFrames(df_customers)

^^^^^^^^^^^^^^
Tipo de dato
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Tamaño
(99441, 5)
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Cantidad de nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Recuento de duplicados
0


In [14]:
datosDataFrames(df_order_payments)

^^^^^^^^^^^^^^
Tipo de dato
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Tamaño
(103886, 5)
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Cantidad de nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
^^^^^^^^^^^^^^
^^^^^^^^^^^^^^
Recuento de duplicados
0


# LIMPIEZA DEL DATAFRAME

1) Analizamos el porcentaje de nulos en cada dataFrame, si el resultado no es representativo podemos optar por eliminar dichas filas.

DataFrame: ORDERS

In [15]:
#Se obtiene el porcentaje de nulos en cada columna y visualización
null_percentages_orders = (df_orders.isnull().sum() / len(df_orders)) * 100
print(null_percentages_orders)

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64


In [16]:
#Teniendo en cuenta que menos del 10% no es un valor representativo para el DF
#Filtramos solo las columnas con menos del 10% de valores nulos
columns_to_keep_orders = null_percentages_orders[null_percentages_orders <= 10].index

#Eliminamos las filas con nulos de dichas columnas
df_orders.dropna(subset=columns_to_keep_orders, inplace=True)

#Controlamos la cantidad de nulos para asegurarnos de haberlos eliminado
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

DataFrame: PRODUCTS

In [17]:
#Porcentaje de nulos en cada columna y visualización
null_percentages_products = (df_products.isnull().sum() / len(df_products)) * 100
print(null_percentages_products)

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64


In [18]:
#Teniendo en cuenta que menos del 10% no es un valor representativo para el DF
#Filtramos solo las columnas con menos del 10% de valores nulos
columns_to_keep_orders = null_percentages_products[null_percentages_products <= 10].index

#Eliminamos las filas con nulos de dichas columnas
df_products.dropna(subset=columns_to_keep_orders, inplace=True)

#Controlamos la cantidad de nulos para asegurarnos de haberlos eliminado
df_products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

2) Se transforman los tipos de datos para que todos sean homogeneos

In [19]:
#Se crean 4 funciones para transformar fecha, string, mayúsculas y minúsculas:

#Transforma las columnas de columnas_fechas a datetime
def transformarFechas(df, columnas_fechas):
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime)
    return df

#Transforma las columnas en columnas_string a string
def transformarString(df, columnas_string):
    for columna in columnas_string:
        df[columna] = df[columna].astype("string")
    return df

#Convierte todos los valores de la columna a mayúsculas
def convertirMayusculas(df, columna):
    df[columna] = df[columna].str.upper()
    return df

#Convierte todos los valores de la columna a minúsculas
def convertirMinusculas(df, columna):
    df[columna] = df[columna].str.lower()
    return df

DataFrame: ORDERS

In [20]:
#Se crea una lista de columnas que contienen fechas
columnas_fechas_orders = ['order_purchase_timestamp', 'order_approved_at','order_delivered_carrier_date', 'order_delivered_customer_date',
                          'order_estimated_delivery_date']

#Aplicamos la función que convierte las fechas a datatime
df_orders = transformarFechas(df_orders, columnas_fechas_orders)

In [21]:
#Se crea una lista de columnas a transfomar en string
columnas_string_orders = ["order_id","customer_id","order_status"]

#Aplicamos la función que convierte de object a string
df_orders = transformarString(df_orders, columnas_string_orders)

In [22]:
#Aplicamos la función que convierte a minúsculas
df_orders = convertirMinusculas(df_orders, "order_status")

In [23]:
#Se corrobora que las transformaciones se hayan realizado correctamente
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB


DataFrame: ORDER_ITEMS

In [24]:
#Se crea una lista de columnas que contienen fechas
columnas_fechas_order_items = ['shipping_limit_date']

#Aplicamos la función que convierte las fechas a datatime
df_order_items = transformarFechas(df_order_items, columnas_fechas_order_items)

In [25]:
#Se crea una lista de columnas a transfomar en string
columnas_string_order_items = ["order_id","order_item_id","product_id","seller_id"]

#Aplicamos la función que convierte de object a string
df_order_items = transformarString(df_order_items, columnas_string_order_items)

In [26]:
#Se corrobora que las transformaciones se hayan realizado correctamente
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  string        
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), string(4)
memory usage: 6.0 MB


DataFrame: PRODUCTS

In [27]:
#Se crea una lista de columnas a transfomar en string
columnas_string_products = ["product_id","product_category_name"]

#Aplicamos la función que convierte de object a string
df_products = transformarString(df_products, columnas_string_products)

In [28]:
#Aplicamos la función que convierte a minúsculas
df_products = convertirMinusculas(df_products, "product_category_name")

In [29]:
#Se corrobora que las transformaciones se hayan realizado correctamente
df_products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB


DataFrame: CUSTOMERS

In [30]:
#Se crea una lista de columnas a transfomar en string
columnas_string_customers = ["customer_id","customer_unique_id","customer_zip_code_prefix","customer_city","customer_state"]

#Aplicamos la función que convierte de object a string
df_customers = transformarString(df_customers, columnas_string_customers)

In [31]:
#Aplicamos la función que convierte a mayúculas
df_customers = convertirMayusculas(df_customers, "customer_state")

In [32]:
#Aplicamos la función que convierte a minúsculas
df_customers = convertirMinusculas(df_customers, "customer_city")

In [33]:
#Se corrobora que las transformaciones se hayan realizado correctamente
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


DataFrame: PAYMENTS

In [34]:
#Se crea una lista de columnas a transfomar en string
columnas_string_order_payments = ["order_id","payment_type"]

#Aplicamos la función que convierte de object a string
df_order_payments = transformarString(df_order_payments, columnas_string_order_payments)

In [35]:
#Aplicamos la función que convierte a minúsculas
df_order_payments = convertirMinusculas(df_order_payments, "payment_type")

In [36]:
#Se corrobora que las transformaciones se hayan realizado correctamente
df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB


# ANALISIS DEL DATAFRAME

Es recomendable para poder realizar un análisis ordenado de la información, separar los datos que sean cualitativos de los cuantitativos

In [37]:
#Para ello, creamos una función que separa las columnas cualitativas de las cuantitativas.
#Luego lo aplicamos a cada DataFrame
def analisisDataFrames(df):
  print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
  print("Variables cuantitativas")
  columnas_numericas = df.select_dtypes(include=['int', 'float']).columns
  if not columnas_numericas.empty:
    print(df[columnas_numericas].describe())
  print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
  print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
  print("Variables cualitativas")
  columnas_string = df.select_dtypes(include=['string']).columns
  if not columnas_string.empty:
    print(df[columnas_string].describe())

DataFrame: ORDERS

In [38]:
analisisDataFrames(df_orders)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cuantitativas
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cualitativas
                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       delivered  
freq          96455  


In [39]:
df_orders["order_status"].value_counts()

order_status
delivered    96455
canceled         6
Name: count, dtype: Int64

Observaciones:

1.   De 96.461 pedidos fueron entregados 96.455 es decir el 99,99% de las compras realizadas se resolvieron con éxito.




DataFrame: ORDER_ITEMS

In [40]:
analisisDataFrames(df_order_items)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cuantitativas
               price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cualitativas
                                order_id order_item_id  \
count                             112650        112650   
unique                             98666            21   
top     8272b63d03f5f79c56e9e4120aec44ef             1   
freq                                  21         98666   

                              product_id                         seller_id  
count                             112650                            112650  
unique                             32951                              3095  
top     

Observaciones:

1.   El precio promedio de los productos es de 120.65 reales.
2.   Si bien el valor promedio es de 120.65 se puede apreciar que dicho valor es afectado por algunos precios excesivamente caros (como el precio máximo de 6735.0 reales).
3.   El 50% de las compras han sido por productos de menos de 74.99 reales y el 75% de las compras representan pagos de menos de 134.90 reales lo nos lleva a la conclusión de poder apreciar que la mayoría de las compras son de bajos costos.
4.   Los precios de las entregas promedio rondan en los 19.9 reales pero se puede observar que dicho promedio es influenciado por algunas entregas de grandes costos (como el precio máximo de 409.68 reales). El 50% de las entregas han pagado menos de 16.26 reales y el 75% menos de 21.15 reales lo que nos lleva a concluir que el costo de las entregas no es demasiado caro.



DataFrame: PRODUCTS

In [41]:
analisisDataFrames(df_products)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cuantitativas
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965   

In [42]:
df_products['product_category_name'].value_counts()

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: Int64

Observaciones:


1.   La cantidad total de productos vendidos fueron de 32.340 unidades.
2.   Los productos más vendidos pertenecen a la categoría de artículos del hogar y baño, seguida por deporte y ocio, y en tercer lugar por muebles y decoración.
3.   Los artículos menos vendidos fueron los de la categoría CD's/DVD's musicales junto con seguros y servicios y productos de pc/gamer.



DataFrame: CUSTOMERS

In [43]:
analisisDataFrames(df_customers)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cuantitativas
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cualitativas
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746  


In [44]:
df_customers["customer_city"].value_counts()

customer_city
sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
palmares paulista        1
pires ferreira           1
vicente dutra            1
alagoa nova              1
eugenio de castro        1
Name: count, Length: 4119, dtype: Int64

Observaciones:

1.   La cantidad total de clientes unicos es de 99.441
2.   La mayoria de las compras fueron realizadas desde Sao Pablo seguido de Río de Janeiro y en tecer lugar Belo Horizonte.
3.   Los lugares donde se registraron muy pocas ventas fueron: Palmares paulista, Pires ferreira, Vicente dutra, Alagoa nova y Eugenio de castro.



DataFrame: ORDER_PAYMENTS

In [45]:
analisisDataFrames(df_order_payments)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cuantitativas
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Variables cualitativas
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  credit_card
freq                                  29        76795


In [46]:
df_order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: Int64

Observaciones:

1.   Se realizaron en total 103.886 pagos.
2.   El medio de pago más utilizado por los clientes fue a través de tarjetas de credito, seguido de boletos y en tercer lugar por voucher.



# FUSION DE DATAFRAMES

Se realiza una unión entre dataFrames que comparten datos para simplificar la visualización de la información.

In [47]:
#Se utiliza un merge para la fusión entre el df_orders y df_orders_payments.
#La columna que ambos comparten es order_id
df_merge_orders_payments = pd.merge(df_orders, df_order_payments, how = "inner", on= "order_id")
df_merge_orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0                    2017-10-18                   1  credit_card   
1                    2017-10-18                   3      voucher   
2                    2017-10-18                   2      voucher   
3                    2018-08-13                   1       boleto   
4                    2018-09-04                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [48]:
#Se eliminan aquellas columnas donde la información no es relevante para el análisis visual
df_merge_orders_payments.drop(['customer_id','order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'], axis=1, inplace=True)
df_merge_orders_payments.head()

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
1  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
2  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
3  53cdb2fc8bc7dce0b6741e2150273451    delivered      2018-07-24 20:41:37   
4  47770eb9100c2d0c44946d9cf07ec65d    delivered      2018-08-08 08:38:49   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  credit_card                     1          18.12  
1                   3      voucher                     1           2.00  
2                   2      voucher                     1          18.59  
3                   1       boleto                     1         141.46  
4                   1  credit_card                     3         179.12

In [49]:
#Se utiliza un merge para la fusión entre el df_order_items y df_products.
#La columna que ambos comparten es product_id
df_merge_order_items_products = pd.merge(df_order_items, df_products, how = "inner", on= "product_id")
df_merge_order_items_products.head()

order_id order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214             1   
1  00018f77f2f0320c557190d7a144bdd3             1   
2  000229ec398224ef6ca0657da4fc703e             1   
3  00024acbcdf0a6daa1e931b038114c75             1   
4  00042b26cf59d7ce69dfabb4e55b4fd9             1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value product_category_name  \
0 2017-09-19 09:45:35   58.90          13.29            cool_stuff   
1 2017-05-03 11:05:13  239.90          19.93              pet_shop   
2 2018-01-18 14:48:30  199.00          17.87      moveis_decoracao   
3 2018-08-15 10:10:18   12.99          12.79            perfumaria   
4 2017-02-13 13:57:51  199.90          18.14    ferramentas_jardim   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 56.0                       239.0                 2.0   
2                 59.0                       695.0                 2.0   
3                 42.0                       480.0                 1.0   
4                 59.0                       409.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1           30000.0               50.0               30.0              40.0  
2            3050.0               33.0               13.0              33.0  
3             200.0               16.0               10.0              15.0  
4            3750.0               35.0               40.0              30.0

In [50]:
#Se eliminan aquellas columnas donde la información no es relevante para el análisis visual
df_merge_order_items_products.drop(['product_id','seller_id','shipping_limit_date','product_name_lenght', 'product_description_lenght','product_weight_g', 'product_length_cm', 'product_height_cm','product_width_cm'], axis=1, inplace=True)
df_merge_order_items_products.head()

order_id order_item_id   price  freight_value  \
0  00010242fe8c5a6d1ba2dd792cb16214             1   58.90          13.29   
1  00018f77f2f0320c557190d7a144bdd3             1  239.90          19.93   
2  000229ec398224ef6ca0657da4fc703e             1  199.00          17.87   
3  00024acbcdf0a6daa1e931b038114c75             1   12.99          12.79   
4  00042b26cf59d7ce69dfabb4e55b4fd9             1  199.90          18.14   

  product_category_name  product_photos_qty  
0            cool_stuff                 4.0  
1              pet_shop                 2.0  
2      moveis_decoracao                 2.0  
3            perfumaria                 1.0  
4    ferramentas_jardim                 1.0

# NUEVOS DATAFRAME

Se crea una nueva carpeta para guardar los nuevos dataframe depurados

In [51]:
#Crear la carpeta 'dataset_generados' si no existe
if not os.path.exists('dataset_generados'):
    os.makedirs('dataset_generados')

In [54]:
#Exportamos como csv nuestros nuevos dataframes
df_merge_orders_payments.to_csv("dataset_generados/orders_payments.csv")
df_merge_order_items_products.to_csv("dataset_generados/order_items_products.csv")
df_customers.to_csv("dataset_generados/customers.csv")